src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,209 @@
+                    _certHandle.Dispose();
+                }
+            }
+
+            private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr privateKeyHandle)
+            {
+                using (Interop.libssl.SafeSslHandle sslHandle = new Interop.libssl.SafeSslHandle(ssl))
+                {
+                    certHandle = IntPtr.Zero;
+                    privateKeyHandle = IntPtr.Zero;
+                    VerboseTrace("libssl's client certificate callback");
+                    ISet<string> issuerNames = GetRequestCertificateAuthorities(sslHandle);
+                    X509Certificate2 certificate;
+                    X509Chain chain;
+                    if (!GetClientCertificate(issuerNames, out certificate, out chain))
+                    {
+                        VerboseTrace("no cert or chain");
+                        return 0;
+                    }
+
+                    Debug.Assert(chain != null && chain.ChainElements.Count > 0, " a X509Chain received must not be empty");
+
+                    using (RSAOpenSsl rsa = certificate.GetRSAPrivateKey() as RSAOpenSsl)
+                    using (ECDsaOpenSsl ecdsa = certificate.GetECDsaPrivateKey() as ECDsaOpenSsl)
+                    {
+                        VerboseTrace("rsa = " + rsa);
+                        if (rsa != null)
+                        {
+                            _privateKeyHandle = rsa.DuplicateKeyHandle();
+                        }
+                        else if (ecdsa != null)
+                        {
+                            _privateKeyHandle = ecdsa.DuplicateKeyHandle();
+                        }
+                    }
+
+                    if (_privateKeyHandle == null)
+                    {
+                        VerboseTrace("invalid private key");
+                        return 0;
+                    }
+                    _certHandle = Interop.libcrypto.X509_dup(certificate.Handle);
+
+                    // add extra chain cert
+                    IntPtr sslContext = Interop.libssl.SSL_get_SSL_CTX(sslHandle);
+                    Debug.Assert (sslContext != IntPtr.Zero, "SSL_CTX cannot be null");
+                    using (SafeSslContextHandle contextHandle = SafeSslContextHandle.FromSslContextPtr(sslContext))
+                    {
+                        for (int i = chain.ChainElements.Count - 2; i > 0; i--)
+                        {
+                            SafeX509Handle dupCertHandle = Interop.libcrypto.X509_dup(chain.ChainElements[i].Certificate.Handle);
+                            if (Interop.libcurl.AddExtraChainCertToSslCtx(contextHandle, dupCertHandle) != 1)
+                            {
+                                VerboseTrace("failed to add extra chain cert");
+                                return -1;
+                            }
+                        }
+                    }
+
+                    certHandle = _certHandle.DangerousGetHandle();
```

Comment:
I wasn't sure what the best advice to give here was.

The values being emitted via the out parameters are going to be AddRef'd on the native side.  But we need the runtime here to not free the structure until that happens.

Since the original _privateKeyHandle and _certHandle are being held still we're good unless there's a Dispose that races the assignment.

Thoughts? :)